### ISSUE TITLE
Make common members accessible without matching

### ISSUE THREAD

**Issue by iamim** [domn1995/dunet#120]
In TypeScript, if all union cases have a field it may be accessed without matching. For example:
```typescript
type Animal = 
  | { kind: 'dog', sound: 'woof', ageInDogYears: 12 }
  | { kind: 'cat', sound: 'meow', ageInCatYears: 23 }

const t: Animal = // some value

t.kind // valid, since both have `kind`
t.sound // valid
t.ageInDogYears // error
```

I believe this library would benefit from this pattern when a property name and type are the same in all cases (TS allows the types to be different in which case it creates an anonymous union type for the return value; I guess it could also be implemented but I see it much less useful).

To implement it the base class needs to match on the type and then return this property, for example:
```csharp
// Types.cs
[Union]
partial record Animal {
  partial record Dog(string Sound) : Animal;
  partial record Cat(string Sound) : Animal;
}

// Types.gen.cs
partial record Animal {
  // possibly getter would be better
  public string Sound() => this switch {
    Dog(sound) => sound,
    Cat(sound) => sound
  }
}
```

**Comment by domn1995** [comment]
Hey, I really appreciate you taking the time to contribute your idea!

I am a big fan of how typescript treats union types as well!

What do you see as the benefit of supporting this pattern over just declaring a member in the union record itself. For example:

```cs
[Union]
public partial record HttpResult<T>
{
    public required int StatusCode { get; init; }

    public partial record Ok(T Body);

    public partial record BadRequest(string Message);

    public partial record ServerError;
}
```

**Comment by iamim** [comment]
I must admit I haven't actually used the library yet. I found it in awesome-source-generators list.
Discriminated unions is something I try to replicate in C# quite often and I found that your library is the closest to how I would do it :) It seems like that it's evolved somehow since then.

I prefer making the base class `abstract` and the cases `sealed` as I don't think that an instance of `Animal` would make sense without actually being either `Dog` or `Cat` in my example. If approached through `abstract` base, `StatusCode` (in your example) would have to be extended in each case and it makes the whole thing ugly. I do see exhaustive `Match` as a big benefit of your approach. However, I do believe that `switch` is important to make the library more approachable and idiomatic.

TL;DR: trade offs :) I can see how your approach works and if you'r ok with it then there's no need in my suggestion.

**Comment by domn1995** [comment]
> I prefer making the base class `abstract` and the cases `sealed` as I don't think that an instance of `Animal` would make sense without actually being either `Dog` or `Cat` in my example. If approached through `abstract` base, `StatusCode` (in your example) would have to be extended in each case and it makes the whole thing ugly. I do see exhaustive `Match` as a big benefit of your approach. However, I do believe that `switch` is important to make the library more approachable and idiomatic.

The source generator actually makes union record abstract and its variants sealed behind the scenes for you. :)

> TL;DR: trade offs :) I can see how your approach works and if you'r ok with it then there's no need in my suggestion.

Sounds good. Would love to hear your feedback!

**Comment by iamim** [comment]
I actually didn't know that `partial` can set `abstract`. I tested with this in mind and it does work :)
The common property accessor would make case construction more uniform but I guess it's not a big deal. Consider your example:
```csharp
var one = new HttpResponse<string>.Ok("Hello") { StatusCode = 200 };
var two = new HttpResponse<string>.Ok(200, "Hello");

one.StatusCode;
two.StatusCode; // if all cases have a property named StatusCode of int
```
Feel free to re-open if you find it a useful upgrade :)

Overall, a great library!
